samples/notebooks/image-captioning-gcp/Image Captioning TF 2.0.ipynb

```diff
@@ -0,0 +1,893 @@
+    "        \n",
+    "        len_result = len(result)\n",
+    "        for l in range(len_result):\n",
+    "            temp_att = np.resize(attention_plot[l], (8, 8))\n",
+    "            plt.title(result[l])\n",
+    "            img = plt.imshow(temp_image)\n",
+    "            plt.imshow(temp_att, cmap='gray', alpha=0.6, extent=img.get_extent())\n",
+    "            \n",
+    "            # Save plt to image to access in tensorboard\n",
+    "            buf = BytesIO()\n",
+    "            plt.savefig(buf, format='png')\n",
+    "            buf.seek(0)\n",
+    "            \n",
+    "            final_im = tf.image.decode_png(buf.getvalue(), channels=4)\n",
+    "            final_im = tf.expand_dims(final_im, 0)\n",
+    "            with summary_writer.as_default():\n",
+    "                tf.summary.image(\"attention\", final_im, step=l)\n",
+    "    \n",
+    "    # Select a random image to caption from validation set\n",
+    "    rid = np.random.randint(0, len(img_name_val))\n",
+    "    image = img_name_val[rid]\n",
+    "    real_caption = ' '.join([tokenizer.index_word[i] for i in cap_val[rid] if i not in [0]])\n",
+    "    result, attention_plot = evaluate(image)\n",
+    "    print ('Image:', image)\n",
+    "    print ('Real Caption:', real_caption)\n",
+    "    print ('Prediction Caption:', ' '.join(result))\n",
+    "    plot_attention(image, result, attention_plot)\n",
+    "    \n",
+    "    # Plot attention images on tensorboard\n",
+    "    metadata = {\n",
+    "        'outputs': [{\n",
+    "            'type': 'tensorboard',\n",
+    "            'source': tensorboard_dir,\n",
+    "        }]\n",
+    "    }\n",
+    "    with open('/mlpipeline-ui-metadata.json', 'w') as f:\n",
+    "        json.dump(metadata, f)\n",
+    "    "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "TARGET_IMAGE = 'gcr.io/%s/predict:latest' % PROJECT_NAME\n",
+    "predict_op = compiler.build_python_component(\n",
+    "    component_func=predict,\n",
+    "    staging_gcs_path=PIPELINE_STORAGE_PATH,\n",
+    "    base_image=BASE_IMAGE,\n",
+    "    dependency=[kfp.compiler.VersionedDependency(name='matplotlib', version='3.1.0')],\n",
+    "    target_image=TARGET_IMAGE)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<h3> Create and run pipeline </h3>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<h4> Create pipeline </h4>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@dsl.pipeline(\n",
+    "    name='Image Captioning Pipeline',\n",
+    "    description='A pipeline that trains a model to caption images'\n",
+    ")\n",
+    "def caption_pipeline(\n",
+    "dataset_path=GCS_DATASET_PATH,\n",
```

Comment:
nit: indentation for the code to look more readable.

Reply:
Done